[Fix #7291] Add better message for layout/HashAlignment

jsartin513 opened the pull request:
Fixes [#7291](https://github.com/rubocop-hq/rubocop/issues/7291)

Before this change, the Layout/HashAlignment cop always gave the same error message: "Align the elements of a hash literal if they span more than one line."
This was confusing in situations where a hash appeared to be aligned, but was not according to the configured alignment settings.

This PR changes the error message to reflect which part(s) of a hash must be aligned to pass the cop.

-----------------

Before submitting the PR make sure the following are checked:

* [x] Wrote [good commit messages][1].
* [x] Commit message starts with `[Fix #issue-number]` (if the related issue exists).
* [x] Feature branch is up-to-date with `master` (if not - rebase it).
* [x] Squashed related commits together.
* [x] Added tests.
* [x] Added an entry to the [Changelog](https://github.com/rubocop-hq/rubocop/blob/master/CHANGELOG.md) if the new code introduces user-observable changes. See [changelog entry format](https://github.com/rubocop-hq/rubocop/blob/master/CONTRIBUTING.md#changelog-entry-format).
* [x] The PR relates to *only* one subject with a clear title and description in grammatically correct, complete sentences.
* [x] Run `bundle exec rake default`. It executes all tests and RuboCop for itself, and generates the documentation.

[1]: https://chris.beams.io/posts/git-commit/

buehmann commented on lib/rubocop/cop/layout/hash_alignment.rb:
```diff
@@ -248,10 +248,17 @@ def check_pairs(node)
 
+        def specific_message(format)
+          messages = { KeyAlignment => 'Align the keys of a hash literal if they span more than one line.',
+                       SeparatorAlignment => 'Align the separator of a hash literal if they span more than one line.',
```
Shouldn't this read `separators`? (matching `they`)

buehmann commented on lib/rubocop/cop/layout/hash_alignment.rb:
```diff
@@ -248,10 +248,17 @@ def check_pairs(node)
+        end
+
         def add_offences
           _format, offences = offences_by.min_by { |_, v| v.length }
```
Now that `_format` is no longer unused, better drop the underscore from its name.

jsartin513 commented:
@buehmann I've addressed your feedback, I would appreciate if you could take another look!

I'm also getting an `AppVeyor` check failure, I think this is because of merge conflicts in the changelog. I think it makes sense to fix that when I'm otherwise ready to merge.

buehmann commented on lib/rubocop/cop/layout/hash_alignment.rb:
```diff
@@ -248,10 +248,20 @@ def check_pairs(node)
           add_offences
         end
 
+        def specific_message(format)
```
The `MSG` constant near the top of the class can be removed then, can't it?

buehmann commented on lib/rubocop/cop/layout/hash_alignment.rb:
```diff
@@ -248,10 +248,20 @@ def check_pairs(node)
+                       SeparatorAlignment => 'Align the separators of a hash '\
+                       'literal if they span more than one line.',
+                       TableAlignment => 'Align the keys and values of a hash '\
+                       'literal if they span more than one line.' }
```
This is a constant. I think it would be better to treat it as such, that is, extract it as `MESSAGES` from the method.

And it would be great if you could add spaces in front of the backslashes and indent the second lines of the string literals. We do not have a cop for this yet, but see #6420. Something like this:

```ruby
        MESSAGES = {
          KeyAlignment => 'Align the keys of a hash literal ' \
            'if they span more than one line.',
```

jsartin513 commented:
Thanks @buehmann ! I've addressed your feedback, getting some test failures on CircleCI that seem unrelated (a local `rake default` looks fine, and these errors existed in some previous runs). Any suggestions on debugging that, or should I leave it alone for now?

buehmann commented:
@jsartin513 Thanks! I believe those test failures will be fixed by https://github.com/rubocop-hq/rubocop/pull/7597

jsartin513 commented:
Great, thank you @buehmann . 

Is there anything else I should do to push this forwards, or do you take it from here?

koic commented:
Yeah, #7597 has been merged.

koic commented:
@jsartin513 Can you rebase with the latest master branch?

jsartin513 commented:
@koic Done and all checks are passing now!

koic commented on CHANGELOG.md:
```diff
@@ -11,6 +11,7 @@
 * [#7193](https://github.com/rubocop-hq/rubocop/issues/7193): Prevent `Style/PercentLiteralDelimiters` from changing `%i` literals that contain escaped delimiters. ([@buehmann][])
 * [#7590](https://github.com/rubocop-hq/rubocop/issues/7590): Fix an error for `Layout/SpaceBeforeBlockBraces` when using with `EnforcedStyle: line_count_based` of `Style/BlockDelimiters` cop. ([@koic][])
 * [#7569](https://github.com/rubocop-hq/rubocop/issues/7569): Make `Style/YodaCondition` accept `__FILE__ == $0`. ([@koic][])
+* [#7291](https://github.com/rubocop-hq/rubocop/issues/7291): Fix misleading error messages in `Layout/AlignHash` cop to specify how to align the hash. ([@jsartin513][])
```
I think changelog entry is unnecessary because this update does not change a behavior of the `Layout/HashAlignment` cop feature.

bbatsov commented:
Thanks for tackling this!
